osgar/drivers/logsocket.py

```diff
@@ -49,25 +49,92 @@ def run_input(self):
     def run_output(self):
         try:
             while True:
-                __, __, data = self.bus.listen()
-                self._send(data)
+                __, channel, data = self.bus.listen()
+                if channel == 'raw':
+                    self._send(data)
+                else:
+                    assert False, channel  # unsupported channel
+ 
         except BusShutdownException:
             pass
 
     def request_stop(self):
         self.bus.shutdown()
 
 
-class LogTCP(LogSocket):
+class LogTCPBase(LogSocket):
+    """
+      TCP base class for different use cases
+    """
     def __init__(self, config, bus):
         soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        LogSocket.__init__(self, soc, config, bus)
-        self.socket.connect(self.pair)
+        super().__init__(soc, config, bus)
 
     def _send(self, data):
         self.socket.send(data)
 
 
+class LogTCPStaticIP(LogTCPBase):
+    """
+      TCP driver for existing static IP (i.e. SICK LIDAR)
+    """
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        self.socket.connect(self.pair)
+
+
+class LogTCPDynamicIP(LogTCPBase):
+    """
+      TCP driver for dynamic previously unknown address
+         (ROS proxy for subscribers)
+    """
+    def start(self):
+        # on start the address is unknown - it will be received by "output_thread"
+        self.output_thread.start()
+
+    def join(self, timeout=None):
+        if self.input_thread.is_alive():
+            self.input_thread.join(timeout=timeout)
```

Comment:
I think this part is racy. I suppose `join` will be called from some "main thread" while the `input_thread` will be started from `output_thread` so the test can say false, but the thread could be started later. I am neither yet sure how much this matters nor what the best solution would be (if it does).

Where is `join` called from and when?

Reply:
Oh, the output thread should be joined first so that it cannot start the input thread.

Reply:
OK, thanks - that seems to be the right solution. :+1: